text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,523 @@
+
+| Description                | Raw size    | Base64ed size |
+|----------------------------|-------------|---------------|
+| Plaintext                  | 23k         | -             |
+| gzipped                    | 3.6k        | 4.8k          |
+| Zopfli                     | 3.4k        | 4.6k          |
+| bzip2                      | 3.4k        | 4.6k          |
+| Affix-grouped (plaintext)  | 7.4k        | -             |
+| Affix-grouped gzipped      | 3.0k        | 4.0k          |
+| Bloom Filter (1% error)    | 720b        | 960b          |
+
+> base64: the result of binary compression will need to be base64-encoded to
+> be transmitted via a CloudFormation template.
+
+Affix-grouped, gzipped data seems to give the best results that are still
+convenient and generic to work with (see Appendix B for a description of
+prefix grouping).
+
+See Appendix C for an example blob to be added to a template. I'm not super
+happy with this, but this is the most straightfoward option available.
+
+> If we are willing to do more work, we can achieve less space used up by doing
+> more esoteric things. One option would be to use a Bloom Filter, which can use
+> a configurable amount of storage to achieve a configurable hit error ratio.
+> For example, achieving about a 1% false positive rate would take on the order
+> of about ~1k base64-encoded bytes. However, in order to extract a construct list
+> from this Bloom Filter would be a considerable amount of work: we would need
+> a full list of all construct available in every library and at every version,
+> in order to test each construct's presence in the filter. This imposes a lot
+> of complexity on the receiver, which is probably not worth the space savings.
+
+### Analytics payload
+
+The analytics payload itself will be preceded by an encoding scheme identifier,
+followed by a `:` and the transmitted data:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Analytics: <payload-encoding>:<payload>
+```
+
+The only valid payload encodings at the moment are:
+
+* `deflate64`: base64-encoded, gzipped data.
+* `plain`: plaintext data.
+
+The payload itself is a JSON object with `"constructs"` currently defined as
+the only valid key (but can be extended in the future). This protocol can be
+extended by adding additional keys to the object.
+
+Values inside the JSON object can be annotated with an encoding scheme as
+well. If so, the same key followed by the text `$encoding` is added to the
+same object.
+
+```json
+{
+  "constructs$encoding": "<value-encoding>",
+  "constructs": <construct list data>
+}
+```
+
+The only valid value encodings at the moment are:
+
+* `afx`: string list, affix-encoded as a string (see Appendix B).
+* `literal`: literal JSON value.
+
+The `$encoding` key may be missing, in which case `literal` is implied.
+
+Encoding may change the type (for `afx` the encoded value is a `string` but the decoded
+value is a `string[]`).
+
```

Comment:
One concern this might raise is that the metadata sent to aws will become opaque. As mitigation maybe we can emit an additional file to the cloud assembly (`analytics.json`) with the unencoded value. This way people will have visibility to it if they examine `cdk.out`

Reply:
We can do that, and I'm not opposed to it. I'm not convinced that it'll be necessary though, given that the behavior of the afx-encoded string will be such that if you paste it into `bash` by doing  `echo <afx-encoded-string-here>`, `bash` brace expansion is going to reconstruct the full list.

It's not like it's a secret, we're publishing the spec out in the open. 

But sure, if you feel it should be MORE readable than that, we can also emit the file.

Reply:
I think it will add a level of trust